version: "3.5"

services:
  db:
    image: postgres:16
    hostname: db
    environment:
      - POSTGRES_PASSWORD=rinha
      - POSTGRES_USER=rinha
      - POSTGRES_DB=rinha
    # ports:
    #   - "5432:5432"
    volumes:
      - ./init.sql:/docker-entrypoint-initdb.d/script.sql:ro
    command: postgres -c fsync=off -c full_page_writes=off -c synchronous_commit=off -c log_min_messages=PANIC -c autovacuum=off
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U rinha"]
      interval: 5s
      timeout: 5s
      retries: 5
    deploy:
      resources:
        limits:
          cpus: "0.9"
          memory: "374MB"

  api1: &api
    image: quay.io/allisson/rinha-de-backend-2024-q1-python:latest
    hostname: api1
    environment:
      - rinha_database_url=postgres://rinha:rinha@db:5432/rinha
      - rinha_database_pool_min_size=5
      - rinha_database_pool_max_size=25
    depends_on:
      db:
        condition: service_healthy
    deploy:
      resources:
        limits:
          cpus: "0.2"
          memory: "80MB"

  api2:
    <<: *api
    hostname: api2

  nginx:
    image: nginx:latest
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf:ro
    depends_on:
      - api1
      - api2
    ports:
      - "9999:9999"
    deploy:
      resources:
        limits:
          cpus: "0.2"
          memory: "16MB"

networks:
  default:
    driver: bridge
    name: rinha-nginx-2024q1
